### PR TITLE
Fixes #34526 - Switch Rails version to 6.1

### PR DESCRIPTION
**Pull request by stejskalleos**
**Related PRs:**
* https://github.com/Dynflow/dynflow/pull/411

**Changes**
- Set default Rails version to 6.1
- Required dynfow >= 1.6.5
- Cleanup deprecation warnings in config/as_deprecation_whitelist.yaml
- ActiveSupport::ParameterFilter instead of ActionDispatch::Http::ParameterFilter
- model.errors.attribute_names instead of model.errors.keys
- Fixed issue with iterating over the model.errors
- use .add for adding error messages instead of <<
- 'where.not' use NAND instead of NOR [0]
-

### CODE REVIEW

File: lib/core_extensions.rb
Comment (ezr-ondrej) [inline lib/core_extensions.rb]: Given just ActiveModel is affected, there might be a way to update them.
Could we add this with Deprecation warning and update the occurences in core?

**Comment by adamruzicka** [comment]
Dynflow 1.6.4 with the needed changes is out. Please bump the dependency here

**Comment by jlsherrill** [comment]
[test katello]

File: app/presenters/setting_presenter.rb
Comment (ezr-ondrej) [inline app/presenters/setting_presenter.rb]: I've fixed this in https://github.com/theforeman/foreman/pull/9130 which is better solution than what we have and is backwards compatible.

**Comment by ezr-ondrej** [comment]
the match validation failures are fixed by #9137 for anyone interested, I tried to explain what had gone wrong there in the commit and also in the RM ticket.
The failing tests are passing for me locally after applying that PR onto this PR, so it should unblock this to get in :+1:

File: app/presenters/setting_presenter.rb
Comment (ezr-ondrej) [reply app/presenters/setting_presenter.rb]: It's in, can you rebase pls? :)

File: config/application.rb
Comment (ezr-ondrej) [inline config/application.rb]: Why? 6.1 provides this out of the box? could you do it in separate commit and explain in the commit message why are we doing it? :)

**Comment by stejskalleos** [comment]
Waiting for https://github.com/theforeman/foreman/pull/9137

**Comment by ezr-ondrej** [comment]
There is now #9157 as a minor update, that should go in first and possibly be cherry-picked as this will not get cherry-picked for sure :)

**Comment by ezr-ondrej** [comment]
[test katello]
some fixes were merged there

**Comment by ezr-ondrej** [comment]
The `rescue_from JSON params parse errors` commit has been merged as part of #9157, I think it would be better dropping it, tho @ehelms was glad he could just steal it off ^_^

File: config/settings.yaml.example
Comment (ekohl) [inline config/settings.yaml.example]: I'm wondering: does this PR make it incompatible with Rails 6.0? If so, this should only mention Rails 6.1. If it doesn't, then I'd prefer to first merge a PR that uses the new style (so the setting actually works well) and once packaging is in place we merge the 6.1 change. (My guess is that changing from a hash to an array is incompatible and thus you should explicitly choose one or the other.)

File: test/unit/tasks/reports_test.rb
Comment (ekohl) [inline test/unit/tasks/reports_test.rb]: Is there something tracking this? I think normally we use https://github.com/theforeman/foreman/blob/develop/config/as_deprecation_whitelist.yaml so take care of this.

Speaking of which, I think there's one at least 2 entries there that should be removed as part of this PR.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: test/unit/tasks/reports_test.rb
Comment (stejskalleos) [reply test/unit/tasks/reports_test.rb]: Huh I didn't know about the `as_deprecation_whitelist.yaml`, I'll update the file then.

File: config/settings.yaml.example
Comment (stejskalleos) [reply config/settings.yaml.example]: I don't think it's backward compatible, I'll remove the `6.0`

**Review by ezr-ondrej — DISMISSED** [review]
I believe this is ready. I'm missing the information about the necessary changes in the commit message, but that is me nitpicking, tho if you get around to it, would be lovely to include.

As discussed, we will wait after 3.3 branching, and do it asap after it. Until then we have time to take care of the packaging of it. Please ping me if I do not get around to merging it after branching myself :)

**Comment by ezr-ondrej** [comment]
Btw the deprecation warning in foreman CI can probably be fixed by bump of the dynflow?

**Comment by stejskalleos** [comment]
> Btw the deprecation warning in foreman CI can probably be fixed by bump of the dynflow?

I need to update the `as_deprecation_whitelist.yaml` but I can't reproduce the deprecation warning on the localhost so I needed CI run. After that it will be green again

**Comment by stejskalleos** [comment]
@ekohl @ezr-ondrej And it's :green_circle: I also updated commit message with list of all changes done in the PR

**Review by ezr-ondrej — DISMISSED** [review]
Awesome work on this @stejskalleos ! :+1:

**Comment by ehelms** [comment]
I have staged up the packaging for this change when we are ready -- https://github.com/theforeman/foreman-packaging/pull/7744

**Review by ekohl — DISMISSED** [review]

**Comment by stejskalleos** [comment]
Rebased and fixed task `audits:anonymize`

**Comment by adamruzicka** [comment]
In my rather slim testing environment this seems to work well, but what about katello?

**Comment by ezr-ondrej** [comment]
> In my rather slim testing environment this seems to work well, but what about katello?

I've also tested this and it works fine for me as well. I think Katello is working on fixes, but I think it should be easy fixes and thus I'd coordinate with Katello team so they are aware, but I lean towards merge after branching which should give Katello enough time to fix anything that needs fixing, right?

**Comment by jeremylenz** [comment]
[test katello]

**Review by adamruzicka — APPROVED** [review]
Alright, the changes look sensible, let's get this in

**Comment by jeremylenz** [comment]
[test katello]

**Comment by ehelms** [comment]
There are two things I think we should consider before merging:

 1) Ensure @upadhyeammit as 3.3 release owner is OK with current state of RC process to merge a change like this into development branch (as it could cause some churn)
 2) Get https://github.com/theforeman/foreman-packaging/pull/7744 up to date and as close to merge-able as possible, and that @osousa is OK with the current state of 3.3 packaging to merge these changes into develop

I will work on getting that packaging PR I started up to date

**Comment by ehelms** [comment]
From a packaging standpoint we are ready for this change to go in. I see Katello tests are reporting broken so once that is reconciled ACK from packaging to move forward.

**Comment by upadhyeammit** [comment]
Considering where we are with Foreman 3.3 release ACK to merge this in develop branch.

**Comment by parthaa** [comment]
Created a 2nd PR  for katello -> https://github.com/Katello/katello/pull/10125

Hopefully this sorts out the katello build errors

I am not 100% sure all tests are fixed. Would be nice if there is a way to test this pr with the katello PR.

**Comment by parthaa** [comment]
[test katello]

**Comment by evgeni** [comment]
Green! Thanks Partha!

**Review by jeremylenz — APPROVED** [review]

**Comment by jeremylenz** [comment]
@ehelms can we get a packaging ack please? :)

**Review by ehelms — APPROVED** [review]

**Comment by adamruzicka** [comment]
Thank you @stejskalleos & everyone involved in reviews and related PRs in other repos :tada:

**Pull request merged**
